test/test_server_selection.py

```diff
@@ -32,5 +36,79 @@
 class TestAllScenarios(create_selection_tests(_TEST_PATH)):
     pass
 
+
+class TestCustomServerSelectorFunction(IntegrationTest):
+    def test_functional_select_max_port_number_host(self):
+        # Selector that returns server with highest port number.
+        def custom_selector(selection):
+            if not selection:
+                return selection.with_server_descriptions([])
+            ports = [
+                s.address[1] for s in selection.server_descriptions
+            ]
+            idx = ports.index(max(ports))
+            return selection.with_server_descriptions(
+                [selection.server_descriptions[idx]]
+            )
+
+        # Initialize client with appropriate listeners.
+        listener = EventListener()
+        client = rs_or_single_client(
+            serverSelector=custom_selector,
+            event_listeners=[listener]
+        )
+        self.addCleanup(client.close)
+        coll = client.get_database(
+            'testdb', read_preference=ReadPreference.NEAREST).coll
+        self.addCleanup(client.drop_database, 'testdb')
+
+        # Hack to wait long enough for server to populate node list.
```

Comment:
@ShaneHarvey @behackett I notice we are `sleep`ing elsewhere in the test suite as well. Does that mean doing so is the only way to ensure that the node list is populated properly? If so, how can one optimize the amount of time we pause execution (assuming we want as small a number here as possible to minimize test suite running time)?

Reply:
You could add a [TopologyListener](http://api.mongodb.com/python/3.7.1/api/pymongo/monitoring.html#pymongo.monitoring.TopologyListener) that waits for the topology description, [TopologyDescriptionChangedEvent.new_description.known_servers](http://api.mongodb.com/python/3.7.1/api/pymongo/monitoring.html#pymongo.monitoring.TopologyDescriptionChangedEvent.new_description),  to include all known servers ([isMaster.hosts](https://docs.mongodb.com/manual/reference/command/isMaster/#isMaster.hosts)). The topology listener could then notify the test via a [condition variable](https://docs.python.org/2/library/threading.html#condition-objects).